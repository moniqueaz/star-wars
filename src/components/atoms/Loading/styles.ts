import styled, { css, DefaultTheme } from 'styled-components';

type SizeProps = { size: 'small' | 'medium' | 'large'}

type ColorProps = { color: 'dark' | 'light'}

const wrapperModifiers = {
  small: () => css`
    width: 1.6rem;
    height: 1.6rem;
    border-width: 0.2rem;
    &:before {
      width: 0.2rem;
      height: 0.2rem;
    }
    &:after {
      width: 0.2rem;
      height: 0.2rem;
    }
  `,
  medium: () => css`
    width: 3.2rem;
    height: 3.2rem;
    border-width: 0.3rem;
    &:before {
      width: 0.3rem;
      height: 0.3rem;
    }
    &:after {
      width: 0.3rem;
      height: 0.3rem;
    }
  `,
  large: () => css`
    width: 5.6rem;
    height: 5.6rem;
    border-width: 0.4rem;
    &:before {
      width: 0.4rem;
      height: 0.4rem;
    }
    &:after {
      width: 0.4rem;
      height: 0.4rem;
    }
  `,
  dark: (theme: DefaultTheme) => theme.colors.darkGray,
  light: (theme: DefaultTheme) => theme.colors.lightGray,
};

export const Container = styled.div<SizeProps>`
  ${( { size } ) => css`
    ${wrapperModifiers[size]()};
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  `}
`;

export const Wrapper = styled.div<SizeProps>`
  ${( { size } ) => css`
    display: inline-block;
    position: relative;
    ${wrapperModifiers[size]()};
    border: 0;
  `}
`;

export const RoundedEdge = styled.div<SizeProps & ColorProps>`
  ${( { theme, size, color } ) => css`
    ${wrapperModifiers[size]()}
    border-radius: 50%;
    position: absolute;
    z-index: 2;
    transform: rotate(45deg);
    animation: dotte 0.8s linear infinite;
    animation-delay: -0.45s;
    @keyframes dotte {
      0% {
        transform: rotate(45deg);
      }
      100% {
        transform: rotate(405deg);
      }
    }
    &:before {
      content: '';
      background-color: ${wrapperModifiers[color](theme)};
      border-radius: 50%;
      position: absolute;
      top: 50%;
      left: 0;
      transform: translateY(-50%);
    }

    &:after {
      content: '';
      background-color: ${wrapperModifiers[color](theme)};
      border-radius: 50%;
      position: absolute;
      top: 0;
      left: 50%;
      transform: translateX(-50%);
    }
  `}
`;

export const Bar = styled.div<SizeProps & ColorProps>`
  ${( { theme, size, color } ) => css`
    box-sizing: border-box;
    display: block;
    position: absolute;
    border-radius: 50%;
    border-color: ${wrapperModifiers[color](theme)} transparent transparent transparent;
    border-style: solid;
    animation: spinner 0.8s linear infinite;
    ${wrapperModifiers[size]()};

    &:nth-child(1) {
      animation-delay: -0.45s;
    }
    @keyframes spinner {
      0% {
        transform: rotate(0deg);
      }
      100% {
        transform: rotate(360deg);
      }
    }
  `}
`;
