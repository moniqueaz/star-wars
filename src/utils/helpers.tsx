export const fetcher = (url: string) => fetch(url).then(res => res.json());

export const getRandomInteger = (min: number, max: number) => {
  return Math.floor(Math.random() * (max - min)) + min;
};
