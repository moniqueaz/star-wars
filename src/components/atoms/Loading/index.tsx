import * as S from './styles';

type LoadingProps = {
  size?: 'small' | 'medium' | 'large',
  color?: 'dark' | 'light'
}
const Loading = ( { size = 'medium', color='light' }: LoadingProps) => (
  <S.Container size={size}>
    <S.Wrapper size={size}>
      <S.Bar size={size} color={color} />
      <S.RoundedEdge size={size} color={color} />
    </S.Wrapper>
  </S.Container>
);

export default Loading;
