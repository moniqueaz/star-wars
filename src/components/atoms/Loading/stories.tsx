import { Story, Meta } from '@storybook/react/types-6-0';
import Loading from '.';

export default {
  title: 'Atoms/Loading',
  component: Loading,
} as Meta;

export const Default: Story = args => {
  return <div style={{ position: 'relative', width: '300px', height: '300px', backgroundColor: 'black' }}>
    <Loading {...args}/>
  </div>
  ;
};

export const Dark: Story = () => {
  return <div style={{ position: 'relative', width: '300px', height: '300px', backgroundColor: 'black' }}>
    <Loading color="dark"/>
  </div>
  ;
};

Dark.parameters = {
  backgrounds: {
    default: 'light',
  },
};

export const Small: Story = args => {
  return <div style={{ position: 'relative', width: '300px', height: '300px', backgroundColor: 'black' }}>
    <Loading size="small"/>
  </div>
  ;
};

export const Medium: Story = args => {
  return <div style={{ position: 'relative', width: '300px', height: '300px', backgroundColor: 'black' }}>
    <Loading size="medium"/>
  </div>
  ;
};

export const Larger: Story = args => {
  return <div style={{ position: 'relative', width: '300px', height: '300px', backgroundColor: 'black' }}>
    <Loading size="large"/>
  </div>
  ;
};
