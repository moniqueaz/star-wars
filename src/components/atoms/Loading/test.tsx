import { renderWithTheme } from 'utils/tests/helpers';

import Loading from '.';

describe('<Loading />', () => {
  it('should render', () => {
    const { container } = renderWithTheme(<Loading />);
    expect(container.querySelector('div')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  } );

  it('should render small', () => {
    const { container } = renderWithTheme(<Loading size="small" color="dark"/>);
    expect(container.querySelector('div')).toHaveStyle( { width: '1.6rem' } );
  } );

  it('should render medium', () => {
    const { container } = renderWithTheme(<Loading size="medium" />);
    expect(container.querySelector('div')).toHaveStyle( { width: '3.2rem' } );
  } );

  it('should render large', () => {
    const { container } = renderWithTheme(<Loading size="large" />);
    expect(container.querySelector('div')).toHaveStyle( { width: '5.6rem' } );
  } );
} );
