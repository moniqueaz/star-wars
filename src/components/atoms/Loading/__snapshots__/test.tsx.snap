// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render 1`] = `
.c0 {
  width: 3.2rem;
  height: 3.2rem;
  border-width: 0.3rem;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c0:before {
  width: 0.3rem;
  height: 0.3rem;
}

.c0:after {
  width: 0.3rem;
  height: 0.3rem;
}

.c1 {
  display: inline-block;
  position: relative;
  width: 3.2rem;
  height: 3.2rem;
  border-width: 0.3rem;
  border: 0;
}

.c1:before {
  width: 0.3rem;
  height: 0.3rem;
}

.c1:after {
  width: 0.3rem;
  height: 0.3rem;
}

.c3 {
  width: 3.2rem;
  height: 3.2rem;
  border-width: 0.3rem;
  border-radius: 50%;
  position: absolute;
  z-index: 2;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  -webkit-animation: dotte 0.8s linear infinite;
  animation: dotte 0.8s linear infinite;
  -webkit-animation-delay: -0.45s;
  animation-delay: -0.45s;
}

.c3:before {
  width: 0.3rem;
  height: 0.3rem;
}

.c3:after {
  width: 0.3rem;
  height: 0.3rem;
}

.c3:before {
  content: '';
  background-color: #EAEAEA;
  border-radius: 50%;
  position: absolute;
  top: 50%;
  left: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c3:after {
  content: '';
  background-color: #EAEAEA;
  border-radius: 50%;
  position: absolute;
  top: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c2 {
  box-sizing: border-box;
  display: block;
  position: absolute;
  border-radius: 50%;
  border-color: #EAEAEA transparent transparent transparent;
  border-style: solid;
  -webkit-animation: spinner 0.8s linear infinite;
  animation: spinner 0.8s linear infinite;
  width: 3.2rem;
  height: 3.2rem;
  border-width: 0.3rem;
}

.c2:before {
  width: 0.3rem;
  height: 0.3rem;
}

.c2:after {
  width: 0.3rem;
  height: 0.3rem;
}

.c2:nth-child(1) {
  -webkit-animation-delay: -0.45s;
  animation-delay: -0.45s;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      color="light"
    />
    <div
      class="c3"
      color="light"
    />
  </div>
</div>
`;
